### DISCUSSION TITLE
Is there a way to view the number of cells in an `.h5m` file?

### DISCUSSION THREAD

**Discussion by jbae11** [svalinn/DAGMC#811] [General]
Is there a convenient way to 'inspect' an `.h5m` file? For example printing out how many cells it has?

Thank you.

**Comment by makeclean** [comment]
I recall someone writing an h5m file inspector @

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Jin Whan Bae ***@***.***>
Sent: Monday, July 4, 2022 7:49:37 PM
To: svalinn/DAGMC ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [svalinn/DAGMC] Is there a way to view the number of cells in an `.h5m` file? (Discussion #811)


Is there a convenient way to 'inspect' an .h5m file? For example printing out how many cells it has?

Thank you.

**Reply by shimwell** [reply]
Here is a link to the dagmc_h5m_file_inspector https://github.com/fusion-energy/dagmc_h5m_file_inspector it was a community effort

**Comment by gonuke** [comment]
We are also developing a DAGMC stats package to generate summaries of many characteristics of a DAGMC file: https://github.com/svalinn/dagmc_stats

**Comment by jbae11** [comment]
thanks! That sounds great.

**Comment by shimwell** [comment]
@jbae11 give this a go for counting the number of volumes

```bash
pip install h5py
```
Then from within python
```python
import h5py
dagmc_file_contents = h5py.File('dagmc.h5m')
groups=dagmc_file_contents['/tstt/sets/children']
print(len(groups))
```

**Reply by pshriwise** [reply]
I think we need to be careful here, children in the `.h5m` file can refer to surfaces, curves, and geometric vertices also -- not just volumes (a.k.a cell as far as the Monte Carlo code is concerned). The above command might be misleading.

Looking at the structure of the `.h5m` file a little more I think something like this would do the job:

```python
import h5py
dagmc_file_contents = h5py.File('dagmc.h5m')
category_data = dagmc_file_contents['tstt/tags/CATEGORY/values']
categories_as_str = [c.tobytes().decode().replace('\x00', '') for c in category_data]
n_vols = sum([v == 'Volume' for v in categories_as_str])
print(n_vols)
```

FWIW @shimwell, I'm liking the use of `h5py` to get this information from the DAGMC file more and more. The risk is of course that the MOAB file format changes at some point, but I don't see that happening anytime soon and we're well connected enough with the developers that I think we'll see it coming.

**Reply by shimwell** [reply]
Just wondering if this volume code would be useful in the openmc.DAGMCUniverse class 👀 
Are the volumes in a DAGMC file always sequential and start at 1? For example volume ids would be 1,2,3,4

**Reply by pshriwise** [reply]
We could in theory look up the relevant ID information from tag data as well… here’s where accessing that information using the MOAB interface (via PyMOAB) becomes useful. Doing this via the MOAB library ensures we’re correlating data entries in the tag info correctly for a given entity. It would take some manual book keeping to do this with `h5py` I think. Looking forward to discussing this further in our bi-weekly meeting.

**Comment by makeclean** [comment]
Volumes id's are arbitrary and gappy
________________________________

**Reply by shimwell** [reply]
thanks
